### ISSUE TITLE
Improvements to draggers UI

### ISSUE THREAD

**Issue by Animal42069** [Mantas-2155X/AI_BetterHScenes#3]
Great Mod, just one suggestion that would immensely improve QoL.  Due to using adjustable player height, I often find myself using the draggers UI to get everyone into the right position.  It would be awesome if those offsets could be saved off so that the next time that position with those characters is loaded those offsets would be automatically applied.

I have 10+ years of experience with Visual Studio, C++ and C#, so I'm willing to contribute to this.  I would just need to bring myself up to speed with Unity and the Mod tools in general.

**Comment by Mantas-2155X** [comment]
Hello,

I was thinking of making the changes persist but it's too low on my priorities list at the moment. Feel free to contribute! I'd love to see what results you can pull off :)

**Comment by Animal42069** [comment]
I'll give it a shot, it looks like it will be a little harder than I
thought at first because it looks like the UI coordinate system isn't the
same as the character coordinate system, so an X/Z adjustment won't put the
character in the same position every time.  I haven't looked at any code
yet to see what is possible

**Comment by Animal42069** [comment]
Posting an update in case anyone else sees it to offer some advice.

I have most of the changes completed.  I have adjusted the draggers offsets to be relative to character coordinates rather than world coordinates.  They're being saved and recalled to an xml file.

The one piece I am struggling with is that I am applying the offsets during ChangeAnimation, it appears the offsets get applied before the animation change is complete, so that they end up getting reset to 0 when the animation finally loads.  I need to figure out a way to apply the offsets after the animation change is complete, and all else should be good.

I tried putting it in a postfix hoping that would make it run after the animation changed, but that did not work.
        [HarmonyPostfix, HarmonyPatch(typeof(HScene), "ChangeAnimation")]
        private static void HScene_ChangeAnimationCompleted()

My quick and dirty workaround is that I have added a hotkey to load the offsets for the current animation.  This works, and if that's what I have to settle for then so be it, it is still a huge QoL improvement being able to make all tweaks with one key.  But I would really like to figure out a way to make it automatically apply.

**Comment by Animal42069** [comment]
It looks like monitoring the state of hScene.NowChangeAnim and waiting until it transitions to false before applying offsets works.

I might try adding that kind of check to the PVCollider project to see if it removes some of the bugs.

**Comment by Mantas-2155X** [comment]
Hello again.
Thanks for your great work!
The reason why that postfix didn't work is because it's an IEnumerator and only prefixes work on them the "easy" way. Postfixes need to be done slightly differently.
If possible, fork the current repository and push your changes so I, or anyone else can see them.
If you think that you've done enough, feel free to open a pull request :)

**Comment by Animal42069** [comment]
I need to do some more testing and add in a little bit more exception
handling.  Not a lot of free time today but I'll see if I can get the
branch up late tonight

**Comment by Animal42069** [comment]
FYI I had to create a fork, I didn't seem to be able to create a branch.  I don't know if it was user error or I just didn't have permissions to do so.  Sorry for all of the commits in the fork, I'm used to using TFS, git is still new to me.  Everything should be in order now.  So far testing has been going well

**Comment by Mantas-2155X** [comment]
I've created a new [repo](https://github.com/Mantas-2155X/BetterHScenes) for BetterHScenes which also includes HS2. (this one will be archived later)
If possible, could you fork that and do your changes (to the AI version)? We can also create this Issue on the other repo.
Also, do you think just assigning the data to the chara cards themselves with extsave would be a better option than having a separate xml?

**Comment by Animal42069** [comment]
The downside to saving offsets on the character card is that they might not be correct all the time.  The xml file saves the offsets for a given pair of characters.  So for example if you had an HScene with male1/female1 and you saved male1 offsets to the card, those offsets may not be appropriate for a male1/female2 hscene.  I think the long term goal would be to look at the height of each character and use an algorithm to determine offsets, but that's way down the road.

I tried make the offset file in a way that it would work with more than just 2 characters in an HScene, but I haven't tested that out.

I definitely don't like the bloat of the xml file, especially with the commit that I just made.  It allows unique offsets for each of the motions of a given position.  I might try to see if I can just save off values that aren't 0.
